extensions/amp-next-page/1.0/service.js

```diff
@@ -604,6 +604,28 @@ export class NextPageService {
         }
       );
 
+      // Reuse message deliverer from existing viewer.
+      // Even though we have multiple instances of the Viewer service, we only
+      // have a single messaging channel.
+      const messageDeliverer = this.viewer_.maybeGetMessageDeliverer();
+      if (messageDeliverer) {
+        amp.onMessage((eventType, data, awaitResponse) => {
+          // Some messages should be suppressed when coming from the inserted
+          // document. These are related to the viewport, so we allow the host
+          // document to handle these events instead.
```

Comment:
Do they automatically propagate to the host document? Is this early return all that is needed?

Reply:
That's correct, the host document handles the relevant messages properly. Services are instantiated for every document inside the Shadow root, so we're preventing duplication by multiple actors, like the second Viewport service.